File: packages/core/TransmuteCore/src/TransmuteCore/world/World.java
package TransmuteCore.world;

import TransmuteCore.core.Manager;
import TransmuteCore.core.interfaces.services.IRenderer;
import TransmuteCore.graphics.Context;
import TransmuteCore.math.Collision;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

/**
 * Deep tile-world module: grid solids + actors behind one interface.
 * Prefer this over inventing per-game TileMap/Entity classes.
 */
public final class World
{
    public static final int EMPTY = 0;
    public static final int SOLID = 1;

    private final int cols;
    private final int rows;
    private final int tileSize;
    private final int[] tiles;
    private final List<Actor> actors = new ArrayList<>();
    private int clearColor;
    private int solidColor;

    private World(int cols, int rows, int tileSize)
    {
        if (cols <= 0 || rows <= 0 || tileSize <= 0)
        {
            throw new IllegalArgumentException(
                String.format("cols, rows, tileSize must be positive. Got %dx%d tile=%d", cols, rows, tileSize));
        }
        this.cols = cols;
        this.rows = rows;
        this.tileSize = tileSize;
        this.tiles = new int[cols * rows];
        this.clearColor = 0xFF14141E;
        this.solidColor = 0xFF3C3C50;
    }

    public static World grid(int cols, int rows, int tileSize)
    {
        return new World(cols, rows, tileSize);
    }

    public World clearColor(int argb)
    {
        this.clearColor = argb;
        return this;
    }

    public World solidColor(int argb)
    {
        this.solidColor = argb;
        return this;
    }

    public void fillBorder(int tile)
    {
        for (int x = 0; x < cols; x++)
        {
            setTile(x, 0, tile);
            setTile(x, rows - 1, tile);
        }
        for (int y = 0; y < rows; y++)
        {
            setTile(0, y, tile);
            setTile(cols - 1, y, tile);
        }
    }

    public void setTile(int tx, int ty, int tile)
    {
        if (!inBounds(tx, ty))
        {
            throw new IllegalArgumentException(String.format("Tile out of bounds: (%d,%d)", tx, ty));
        }
        tiles[tx + ty * cols] = tile;
    }

    public int getTile(int tx, int ty)
    {
        if (!inBounds(tx, ty))
        {
            return SOLID;
        }
        return tiles[tx + ty * cols];
    }

    public boolean isSolid(int tx, int ty)
    {
        return getTile(tx, ty) == SOLID;
    }

    /**
     * @return true if an axis-aligned box at pixel coords would overlap any solid tile.
     */
    public boolean blocks(int x, int y, int w, int h)
    {
        int minTx = Math.floorDiv(x, tileSize);
        int maxTx = Math.floorDiv(x + w - 1, tileSize);
        int minTy = Math.floorDiv(y, tileSize);
        int maxTy = Math.floorDiv(y + h - 1, tileSize);
        for (int tx = minTx; tx <= maxTx; tx++)
        {
            for (int ty = minTy; ty <= maxTy; ty++)
            {
                if (isSolid(tx, ty))
                {
                    return true;
                }
            }
        }
        return false;
    }

    public void add(Actor actor)
    {
        actors.add(Objects.requireNonNull(actor, "actor"));
        actor.attach(this);
    }

    public List<Actor> getActors()
    {
        return Collections.unmodifiableList(actors);
    }

    public void update(Manager manager, double delta)
    {
        for (Actor actor : actors)
        {
            actor.update(manager, delta);
        }
    }

    public void render(Manager manager, IRenderer renderer)
    {
        Context ctx = (Context) renderer;
        ctx.renderFilledRectangle(0, 0, ctx.getWidth(), ctx.getHeight(), clearColor);
        for (int ty = 0; ty < rows; ty++)
        {
            for (int tx = 0; tx < cols; tx++)
            {
                if (isSolid(tx, ty))
                {
                    ctx.renderFilledRectangle(
                        tx * tileSize, ty * tileSize, tileSize, tileSize, solidColor);
                }
            }
        }
        for (Actor actor : actors)
        {
            actor.render(manager, renderer);
        }
    }

    public int getCols()
    {
        return cols;
    }

    public int getRows()
    {
        return rows;
    }

    public int getTileSize()
    {
        return tileSize;
    }

    public int pixelWidth()
    {
        return cols * tileSize;
    }

    public int pixelHeight()
    {
        return rows * tileSize;
    }

    private boolean inBounds(int tx, int ty)
    {
        return tx >= 0 && ty >= 0 && tx < cols && ty < rows;
    }

    /**
     * Movable thing in a {@link World}. Subclass or use {@link #colored(int, int, int, int, int)}.
     */
    public static class Actor
    {
        protected int x;
        protected int y;
        protected int width;
        protected int height;
        protected int color;
        protected World world;

        public Actor(int x, int y, int width, int height, int color)
        {
            this.x = x;
            this.y = y;
            this.width = width;
            this.height = height;
            this.color = color;
        }

        public static Actor colored(int x, int y, int width, int height, int color)
        {
            return new Actor(x, y, width, height, color);
        }

        void attach(World world)
        {
            this.world = world;
        }

        public void update(Manager manager, double delta)
        {
        }

        public void render(Manager manager, IRenderer renderer)
        {
            Context ctx = (Context) renderer;
            ctx.renderFilledRectangle(x, y, width, height, color);
        }

        /**
         * Tries to move by {@code dx},{@code dy}; rejects the move if any solid blocks the new box.
         *
         * @return true if the move applied
         */
        public boolean tryMove(int dx, int dy)
        {
            if (world == null)
            {
                x += dx;
                y += dy;
                return true;
            }
            int nx = x + dx;
            int ny = y + dy;
            if (world.blocks(nx, ny, width, height))
            {
                return false;
            }
            x = nx;
            y = ny;
            return true;
        }

        public boolean overlaps(Actor other)
        {
            return Collision.aabb(x, y, width, height, other.x, other.y, other.width, other.height);
        }

        public int getX()
        {
            return x;
        }

        public int getY()
        {
            return y;
        }

        public int getWidth()
        {
            return width;
        }

        public int getHeight()
        {
            return height;
        }

        public int getColor()
        {
            return color;
        }

        public World getWorld()
        {
            return world;
        }
    }
}

File: packages/python/src/transmute/world.py
from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path
from typing import Callable


def to_pixel_int(r: int, g: int, b: int, a: int = 255) -> int:
    return ((a & 255) << 24) | ((r & 255) << 16) | ((g & 255) << 8) | (b & 255)


@dataclass
class Actor:
    x: int
    y: int
    width: int
    height: int
    color: int
    name: str | None = None
    world: World | None = None

    def named(self, name: str) -> Actor:
        self.name = name
        return self

    def try_move(self, dx: int, dy: int) -> bool:
        if self.world is None:
            self.x += dx
            self.y += dy
            return True
        nx, ny = self.x + dx, self.y + dy
        if self.world.blocks(nx, ny, self.width, self.height):
            return False
        self.x, self.y = nx, ny
        return True

    def update(self, input_handler=None) -> None:
        pass


@dataclass
class Trigger:
    x: int
    y: int
    width: int
    height: int
    on_enter: Callable[[Actor], None] = field(default_factory=lambda: (lambda a: None))
    id: str | None = None
    _inside: list[Actor] = field(default_factory=list)

    def set_on_enter(self, fn: Callable[[Actor], None]) -> None:
        prev = self.on_enter

        def chained(actor: Actor) -> None:
            prev(actor)
            fn(actor)

        self.on_enter = chained

    def evaluate(self, actors: list[Actor]) -> None:
        now: list[Actor] = []
        for actor in actors:
            if self._overlaps(actor):
                now.append(actor)
                if actor not in self._inside:
                    self.on_enter(actor)
        self._inside = now

    def _overlaps(self, actor: Actor) -> bool:
        return (
            self.x < actor.x + actor.width
            and self.x + self.width > actor.x
            and self.y < actor.y + actor.height
            and self.y + self.height > actor.y
        )


class World:
    EMPTY = 0
    SOLID = 1

    def __init__(self, cols: int, rows: int, tile_size: int = 16) -> None:
        if cols <= 0 or rows <= 0 or tile_size <= 0:
            raise ValueError("cols, rows, tile_size must be positive")
        self.cols = cols
        self.rows = rows
        self.tile_size = tile_size
        self.tiles = [self.EMPTY] * (cols * rows)
        self.actors: list[Actor] = []
        self.triggers: list[Trigger] = []
        self.clear_color = to_pixel_int(20, 20, 30)
        self.solid_color = to_pixel_int(60, 60, 80)

    @classmethod
    def grid(cls, cols: int, rows: int, tile_size: int = 16) -> World:
        return cls(cols, rows, tile_size)

    def with_clear_color(self, argb: int) -> World:
        self.clear_color = argb
        return self

    def with_solid_color(self, argb: int) -> World:
        self.solid_color = argb
        return self

    def fill_border(self, tile: int = SOLID) -> None:
        for x in range(self.cols):
            self.set_tile(x, 0, tile)
            self.set_tile(x, self.rows - 1, tile)
        for y in range(self.rows):
            self.set_tile(0, y, tile)
            self.set_tile(self.cols - 1, y, tile)

    def set_tile(self, tx: int, ty: int, tile: int) -> None:
        if not self._in_bounds(tx, ty):
            raise ValueError(f"Tile out of bounds: ({tx},{ty})")
        self.tiles[tx + ty * self.cols] = tile

    def get_tile(self, tx: int, ty: int) -> int:
        if not self._in_bounds(tx, ty):
            return self.SOLID
        return self.tiles[tx + ty * self.cols]

    def is_solid(self, tx: int, ty: int) -> bool:
        return self.get_tile(tx, ty) == self.SOLID

    def blocks(self, x: int, y: int, w: int, h: int) -> bool:
        min_tx = x // self.tile_size
        max_tx = (x + w - 1) // self.tile_size
        min_ty = y // self.tile_size
        max_ty = (y + h - 1) // self.tile_size
        for tx in range(min_tx, max_tx + 1):
            for ty in range(min_ty, max_ty + 1):
                if self.is_solid(tx, ty):
                    return True
        return False

    def add(self, actor: Actor) -> None:
        self.actors.append(actor)
        actor.world = self

    def remove_actor(self, name: str) -> Actor | None:
        found = self.find_actor(name)
        if found:
            self.actors.remove(found)
            found.world = None
        return found

    def add_trigger(self, trigger: Trigger) -> None:
        self.triggers.append(trigger)

    def find_actor(self, name: str) -> Actor | None:
        for a in self.actors:
            if a.name == name:
                return a
        return None

    def find_trigger(self, id_: str) -> Trigger | None:
        for t in self.triggers:
            if t.id == id_:
                return t
        return None

    def update(self, input_handler=None) -> None:
        for actor in self.actors:
            actor.update(input_handler)
        for trigger in self.triggers:
            trigger.evaluate(self.actors)

    def render(self, ctx) -> None:
        ctx.fill_rect(0, 0, ctx.width, ctx.height, self.clear_color)
        for ty in range(self.rows):
            for tx in range(self.cols):
                if self.is_solid(tx, ty):
                    ctx.fill_rect(
                        tx * self.tile_size,
                        ty * self.tile_size,
                        self.tile_size,
                        self.tile_size,
                        self.solid_color,
                    )
        for actor in self.actors:
            ctx.fill_rect(actor.x, actor.y, actor.width, actor.height, actor.color)

    def pixel_width(self) -> int:
        return self.cols * self.tile_size

    def pixel_height(self) -> int:
        return self.rows * self.tile_size

    def _in_bounds(self, tx: int, ty: int) -> bool:
        return 0 <= tx < self.cols and 0 <= ty < self.rows
